#!/usr/bin/env python
PACKAGE = "thor_mang_ros_control"
def generate(gen):
    #           Name            Type    Level Description         Default  Min   Max
    gen.add("hip_pitch_offset", double_t, 0, "Hip pitch offset", 12.0, 0.0, 20.0)
    gen.add("ankle_pitch_offset", double_t, 0, "Ankle pitch offset", 0.0, -6.0, 6.0)
    gen.add("walk_stabilizer_gain_ratio", double_t, 0, "walk_stabilizer_gain_ratio", 1.5, 0.0, 6.0)
    gen.add("imu_gyro_gain_ratio", double_t, 0, "imu_gyro_gain_ratio", 0.0731, 0.0, 1.0)
    gen.add("force_moment_distribution_ratio", double_t, 0, "imu_gyro_gain_ratio", 0.6, 0.0, 1.0)
    gen.add("balance_hip_pitch_gain", double_t, 0, "balance_hip_pitch_gain", 1.0, 0.0, 3.0)
    gen.add("balance_z_gain_by_ft", double_t, 0, "balance_z_gain_by_ft", 0.05, 0.0, 1.0)
    gen.add("balance_right_roll_gain_by_ft", double_t, 0, "balance_right_roll_gain_by_ft", 0.001, -0.1, 0.1)
    gen.add("balance_right_pitch_gain_by_ft", double_t, 0, "balance_right_pitch_gain_by_ft", -0.0005, -0.1, 0.1)   
    gen.add("balance_left_roll_gain_by_ft", double_t, 0, "balance_left_roll_gain_by_ft", 0.001, -0.1, 0.1)
    gen.add("balance_left_pitch_gain_by_ft", double_t, 0, "balance_left_pitch_gain_by_ft", -0.0005, -0.1, 0.1)
    gen.add("foot_landing_offset_gain", double_t, 0, "balance_left_pitch_gain_by_ft", 0.0, 0.0, 1.0)
    gen.add("foot_landing_detect_n", double_t, 0, "foot_landing_detect_n", 50, 0, 100)
    gen.add("balance_x_gain", double_t, 0, "balance_x_gain", 12,6875, 0, 30)
    gen.add("balance_y_gain", double_t, 0, "balance_y_gain", -15,225, -30, 0)
    gen.add("balance_z_gain", double_t, 0, "balance_z_gain", 0, -100, 100)
    gen.add("balance_pitch_gain", double_t, 0, "balance_pitch_gain", -0,0375, -10, 10)
    gen.add("balance_roll_gain", double_t, 0, "balance_roll_gain", -0,075, -10, 10)
    gen.add("axis_controller_gain", double_t, 0, "axis_controller_gain", 0, -10, 10)
    gen.add("foot_roll_adjustment", double_t, 0, "foot_roll_adjustment", 10, -50, 50)
    gen.add("foot_pitch_adjustment", double_t, 0, "foot_pitch_adjustment", 10, -50, 50)
    gen.add("cob_x_adjustment", double_t, 0, "cob_x_adjustment", 50, -200, 200)
    gen.add("cob_y_adjustment", double_t, 0, "cob_y_adjustment", 50, -200, 200)
    gen.add("p_gain", double_t, 0, "p_gain", 32, 0, 100)
    gen.add("d_gain", double_t, 0, "d_gain", 0, 0, 100)
    gen.add("i_gain", double_t, 0, "i_gain", 0, 0, 100)

    gen.add("custom_imu_gain", double_t, 0, "imu_gyro_gain_ratio", 1, -10, 10.0)

    exit(gen.generate(PACKAGE, "preview_controller_config", "FootstepPreviewController"))

try:
    from dynamic_reconfigure.parameter_generator_catkin import *
    gen = ParameterGenerator()
    generate(gen)
# reason for catching IndexError
# parameter_generator_catkin expects 4 arguments while rosbuild only passes in 2
# not thrilled with this solution
except IndexError:
    print 'ERROR', PACKAGE, 'FootstepPreviewController.cfg failed using parameter_generator_catkin, using rosbuild instead'
    from dynamic_reconfigure.parameter_generator import *
    gen = ParameterGenerator()
    generate(gen)
